### PR TITLE
Consent screen (EXPOSUREAPP-11910)

### PR DESCRIPTION
**Pull request by chiljamgossow**
Screenshot_20220224_101914
Screenshot_20220224_101931
Screenshot_20220224_101944

### CODE REVIEW

**Review by CV113 — DISMISSED** [review]
UA approved

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceBulletPoint.kt [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    init {
+        LayoutInflater.from(context).inflate(R.layout.dcc_reissuance_bullet_point, this, true)
+
+        context.withStyledAttributes(
+            attrs,
+            R.styleable.DccReissuanceBulletPoint,
+        ) {
+            getResourceId(R.styleable.DccReissuanceBulletPoint_android_text, 0).let {
+                bulletPointText.text = if (it != 0) resources.getString(it)
+                else getString(R.styleable.DccReissuanceBulletPoint_android_text)
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceBulletPoint.kt:29]: maybe use `{}` ? for better readability of `if`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceCertificateCard.kt [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import de.rki.coronawarnapp.util.TimeAndDateExtensions.toShortDayFormat
+import de.rki.coronawarnapp.util.TimeAndDateExtensions.toUserTimeZone
+
+class DccReissuanceCertificateCard @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyleAttr: Int = 0
+) : ConstraintLayout(context, attrs, defStyleAttr) {
+
+    val itemView: View = LayoutInflater.from(context).inflate(R.layout.dcc_reissuance_card, this, true)
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceCertificateCard.kt:24]: `private` ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceCertificateCard.kt [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+    val viewBinding: Lazy<DccReissuanceCardBinding> = lazy {
+        DccReissuanceCardBinding.bind(itemView)
+    }
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceCertificateCard.kt:28]: you can use `by` and get rid of `...Binding.value` 

```kotlin
    private val viewBinding by lazy { DccReissuanceCardBinding.bind(itemView) }

```

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceConsentFragment.kt [hunk 4]
```diff
@@ -1,13 +1,71 @@
...
+
+                event.observe2(this@DccReissuanceConsentFragment) {
+                    when (it) {
+                        DccReissuanceConsentViewModel.ReissuanceError -> {
+                            binding.agreeButton.isLoading = false
+                            // show error dialog
+                        }
+                        DccReissuanceConsentViewModel.ReissuanceInProgress -> binding.agreeButton.isLoading = true
+                        DccReissuanceConsentViewModel.ReissuanceSuccess -> {
+                            binding.agreeButton.isLoading = false
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceConsentFragment.kt:61]: Here, code is binding context , you can use the views directly without `binding.`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceConsentFragment.kt [hunk 5]
```diff
@@ -1,13 +1,71 @@
...
     private val viewModel: DccReissuanceConsentViewModel by cwaViewModels { viewModelFactory }
+    private val binding: FragmentDccReissuanceConsentBinding by viewBinding()
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        binding.apply {
+            toolbar.setNavigationOnClickListener { popBackStack() }
+            cancelButton.setOnClickListener { popBackStack() }
+            privacyInformation.setOnClickListener { findNavController().navigate(R.id.informationPrivacyFragment) }
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceConsentFragment.kt:30]: what about using `Event` states here too, to able to unit test these cases ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceConsentFragment.kt [hunk 6]
```diff
@@ -1,13 +1,71 @@
...
+            agreeButton.setOnClickListener { viewModel.startReissuance() }
+
+            viewModel.apply {
+                stateLiveData.observe2(this@DccReissuanceConsentFragment) {
+                    binding.dccReissuanceTitle.text = it.title
+                    binding.dccReissuanceSubtitle.text = it.subtitle
+                    binding.dccReissuanceContent.text = it.content
+                    it.url?.let { url ->
+                        binding.dccReissuanceLink.visibility = View.VISIBLE
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceConsentFragment.kt:39]: to handle all cases , what about `dccReissuanceLink.isVisible = !it.url.isNullOrEmpty()`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceConsentViewModel.kt [hunk 7]
```diff
@@ -1,12 +1,72 @@
...
     interface Factory : SimpleCWAViewModelFactory<DccReissuanceConsentViewModel>
+
+    internal val stateLiveData: MutableLiveData<State> = MutableLiveData()
+    internal val certificateLiveData: MutableLiveData<DccV1.MetaData> = MutableLiveData()
+
+    internal val event = SingleLiveEvent<Event>()
+
+    init {
+        personCertificatesProvider.personCertificates.mapNotNull {
+            val walletInfo = it.first().dccWalletInfo
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceConsentViewModel.kt:34]: this temporary, right ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceConsentViewModel.kt [hunk 8]
```diff
@@ -1,12 +1,72 @@
...
+            }
+
+            walletInfo?.certificateReissuance?.certificateToReissue?.certificateRef?.barcodeData?.let {
+                val certificate = dccQrCodeExtractor.extract(
+                    it,
+                    DccV1Parser.Mode.CERT_SINGLE_STRICT
+                )
+                certificateLiveData.postValue(certificate.data.certificate)
+            }
+        }.launchIn(viewModelScope)
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceConsentViewModel.kt:53]: You can use:
```kotlin
val state = personCertificatesProvider.personCertificates.mapNotNull {
// do mapping here 
  State(
   title = format(division.titleText),
   subtitle = format(division.subtitleText),
   content = format(division.longText),
   url = division.faqAnchor
   )
}.asLiveData2()
```

**Comment by mtwalli** [comment]
Spacing are not adjusted  when `name` is long 
Figma :
Screenshot 2022-02-24 at 12 35 57

PR:
Screenshot 2022-02-24 at 12 46 24

**Comment by mtwalli** [comment]
I think we need different icons for dark them , see other info screens 
Screenshot 2022-02-24 at 12 49 54

**Comment by mtwalli** [comment]
In Turkish legal text the button "label" has to match the actual button label as in other languages , we have got some bugs regarding this before 
Screenshot 2022-02-24 at 12 54 37

See https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-11109 for reference

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceConsentFragment.kt (hunk 5 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceConsentFragment.kt:31]: with unit tests for the view model I cannot cover the navigation, even with ents for these

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceConsentViewModel.kt (hunk 8 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceConsentViewModel.kt:53]: I find it easier the way it is(two separate live data for division and certificate), as division can be set to invisible

File: Corona-Warn-App/src/main/res/values-de/legal_strings.xml [hunk 9]
```diff
@@ -222,4 +222,13 @@
...
 
+    <!-- DCC Reissuance Consent Screen -->
+    <!-- XTXT: DCC Reissuance legal header-->
+    <string name="dcc_reissuance_legal_header" translatable="false">Durch Antippen von „Einverstanden“ willigen Sie in die folgenden Schritte ein:</string>
+    <!-- XTXT: DCC Reissuance legal first bullet point-->
+    <string name="dcc_reissuance_legal_one" translatable="false">Die App übermittelt das Impfzertifikat zur Überprüfung an das RKI. </string>
+    <!-- XTXT: DCC Reissuance legal second bullet point-->
+    <string name="dcc_reissuance_legal_two" translatable="false">Das RKI überprüft das Impfzertifikat und stellt ein aktualisiertes Zertifikat aus und übermittelt dieses an Ihre App.</string>
+    <!-- XTXT: DCC Reissuance button text-->
+    <string name="dcc_reissuance_legal_button" translatable="false">Einverstanden</string>
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/values-de/legal_strings.xml:233]: To be consistent with other consent screens, we should leave the button as it is , but we should change the legal text to be` „Kabul et“ üzerine tıklayarak aş ...`

File: Corona-Warn-App/src/main/res/values-tr/legal_strings.xml [hunk 10]
```diff
@@ -220,4 +220,13 @@
...
 
+    <!-- DCC Reissuance Consent Screen -->
+    <!-- XTXT: DCC Reissuance legal header-->
+    <string name="dcc_reissuance_legal_header" translatable="false">„Onay“ üzerine tıklayarak aşağıdaki adımları kabul edersiniz:</string>
+    <!-- XTXT: DCC Reissuance legal first bullet point-->
+    <string name="dcc_reissuance_legal_one" translatable="false">Uygulama, kontrol amaçlı olarak aşı sertifikasını RKI’ya iletir.</string>
+    <!-- XTXT: DCC Reissuance legal second bullet point-->
+    <string name="dcc_reissuance_legal_two" translatable="false">RKI aşı sertifikasını kontrol eder ve güncel bir sertifika oluşturur ve bunu uygulamanıza aktarır.</string>
+    !-- XTXT: DCC Reissuance button text-->
+    <string name="dcc_reissuance_legal_button" translatable="false">Onay</string>
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/values-tr/legal_strings.xml:231]: To be consistent with other consent screens, we should leave the button as it is , but we should change the legal text to be` „Kabul et“ üzerine tıklayarak aş ...`

**Review by mtwalli — DISMISSED** [review]
LGTM, probably we need another PR for wiring ,but that should not block us from merging. Please consider my comment regarding the legal button label

File: Corona-Warn-App/src/main/res/values-tr/legal_strings.xml (hunk 10 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/res/values-tr/legal_strings.xml:231]: I do not think we can change legal text as it is coming from the lawyers

**Review by CV113 — DISMISSED** [review]
UA approved

File: Corona-Warn-App/src/main/res/values-tr/legal_strings.xml (hunk 10 above)
Comment (mtwalli) [reply Corona-Warn-App/src/main/res/values-tr/legal_strings.xml:231]: You can align with Martin ofc, but last time he said for this case it is fine , see https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-11109?focusedCommentId=1254855&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-1254855

**Review by mtwalli — APPROVED** [review]

File: Corona-Warn-App/src/main/res/values-tr/legal_strings.xml (hunk 10 above)
Comment (mtwalli) [reply Corona-Warn-App/src/main/res/values-tr/legal_strings.xml:231]: will do it in another PR

**Review by AlexanderAlferov — APPROVED** [review]

**Pull request merged**